src/libexpr/parser.y

```diff
@@ -655,13 +660,23 @@ Expr * EvalState::parseExprFromFile(const Path & path)
 
 Expr * EvalState::parseExprFromFile(const Path & path, StaticEnv & staticEnv)
 {
-    return parse(readFile(path).c_str(), foFile, path, dirOf(path), staticEnv);
+    auto buffer = readFile(path);
+    // make sure readFile has left some extra space for terminators
+    assert(buffer.capacity() > 2 && buffer.size() <= buffer.capacity());
+    buffer += '\0';
+    buffer += '\0';
+    return parse(buffer.data(), buffer.size(), foFile, path, dirOf(path), staticEnv);
 }
 
 
 Expr * EvalState::parseExprFromString(std::string_view s, const Path & basePath, StaticEnv & staticEnv)
 {
-    return parse(s.data(), foString, "", basePath, staticEnv);
+    string buffer;
+    buffer.reserve(s.size() + 2);
+    buffer += s;
+    buffer += '\0';
+    buffer += '\0';
```

Comment:
Why do we need two null characters? Seems wasteful to make a copy of the input. If this is really needed, then instead of an `std::string_view`, it should take a `std::string &&` so we don't need to make a copy if the caller doesn't need to keep the string.

Reply:
the generated parser demands them. the function nix previously used for parsing copied the input to add those two, so this is effectively just making behaviour visible that was hidden before.